cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -388,14 +388,14 @@ public void stop(ProgramId programId, @Nullable String runId) throws Exception {
     authorizationEnforcer.enforce(programId, authenticationContext.getPrincipal(), Action.EXECUTE);
     ProgramRuntimeService.RuntimeInfo runtimeInfo = findRuntimeInfo(programId, runId);
     if (runtimeInfo == null) {
-      if (!store.applicationExists(programId.toId().getApplication())) {
-        throw new ApplicationNotFoundException(programId.toId().getApplication());
+      if (!store.applicationExists(programId.getParent().toId())) {
```

Comment:
The Store interface is really large and updating it and its implementations and usages would make this PR much larger.